aeron-driver/src/main/c/aeron_driver_conductor.c

```diff
@@ -705,25 +797,67 @@ aeron_ipc_publication_t *aeron_driver_conductor_get_or_add_ipc_publication(
     aeron_uri_publication_params_t *params,
     int64_t registration_id,
     int32_t stream_id,
-    size_t channel_length,
-    const char *channel,
+    size_t uri_length,
+    const char *uri,
     bool is_exclusive)
 {
     aeron_ipc_publication_t *publication = NULL;
 
-    if (!is_exclusive)
+    uint64_t bits[STATIC_BIT_SET_U64_LEN];
+    aeron_bit_set_t session_id_offsets;
+    aeron_bit_set_stack_init(
+        conductor->network_publications.length + 1, bits, STATIC_BIT_SET_U64_LEN, false, &session_id_offsets);
+    assert(conductor->network_publications.length < session_id_offsets.bit_set_length);
+
+    bool is_session_id_in_use = false;
+
+    for (size_t i = 0; i < conductor->ipc_publications.length; i++)
     {
-        for (size_t i = 0; i < conductor->ipc_publications.length; i++)
-        {
-            aeron_ipc_publication_t *pub_entry = conductor->ipc_publications.array[i].publication;
+        aeron_ipc_publication_t *pub_entry = conductor->ipc_publications.array[i].publication;
 
-            if (stream_id == pub_entry->stream_id &&
-                !pub_entry->is_exclusive &&
-                AERON_IPC_PUBLICATION_STATE_ACTIVE == pub_entry->conductor_fields.state)
+        if (stream_id == pub_entry->stream_id &&
+            pub_entry->conductor_fields.state == AERON_IPC_PUBLICATION_STATE_ACTIVE)
+        {
+            if (NULL == publication && !is_exclusive && !pub_entry->is_exclusive)
             {
                 publication = pub_entry;
-                break;
             }
+
+            if (params->has_session_id && pub_entry->session_id == params->session_id)
+            {
+                is_session_id_in_use = true;
+            }
+
+            aeron_driver_conductor_track_session_id_offsets(conductor, &session_id_offsets, pub_entry->session_id);
+        }
+    }
+
+    int32_t speculated_session_id = 0;
+    int session_id_found = aeron_driver_conductor_speculate_next_session_id(
+            conductor, &session_id_offsets, &speculated_session_id);
+    aeron_bit_set_stack_free(&session_id_offsets);
+
+    if (session_id_found < 0)
+    {
+        aeron_set_err(EINVAL, "(BUG) Unable to allocate session-id");
+        return NULL;
+    }
+
+    if (is_session_id_in_use && (is_exclusive || NULL == publication))
+    {
+        aeron_set_err(
+            EINVAL,
+            "Specified session-id is already in exclusive use for channel: %s, stream-id: %" PRId32,
+            uri, stream_id);
```

Comment:
Need to account for channel length. Example is https://github.com/real-logic/aeron/blob/master/aeron-driver/src/main/c/aeron_position.c#L63